uast/diff/diff.go

```diff
@@ -0,0 +1,283 @@
+package diff
+
+import (
+	// TODO: integrate with https://github.com/src-d/lapjv if perf is unacceptable
+	"fmt"
+	"github.com/heetch/lapjv"
+	"gopkg.in/bblfsh/sdk.v2/uast/nodes"
+)
+
+type decisionType interface {
+	isDecisionType()
+}
+
+type basicDecisionType struct{}
+
+func (_ basicDecisionType) isDecisionType() {}
+
+type decision struct {
+	cost     int
+	decision decisionType
+}
+
+// match decision types together with their params
+type sameDecision struct{ basicDecisionType }
+type replaceDecision struct{ basicDecisionType }
+type matchDecision struct{ basicDecisionType }
+type permuteDecision struct {
+	basicDecisionType
+	permutation []int
+}
+
+//convinience method for choosing the cheapest decision
+func (self *decision) minEq(candidate *decision) {
+	if _, ok := candidate.decision.(replaceDecision); ok || self.cost > candidate.cost {
+		self.cost, self.decision = candidate.cost, candidate.decision
+	}
+}
+
+//type for cache
+type keyType struct{ k1, k2 ID }
+
+//cache for diff computation
+type cacheStorage struct {
+	decisions map[keyType]decision
+	sizes     map[ID]int
+	changes   Changelist
+}
+
+func emptyCacheStorage() cacheStorage {
+	return cacheStorage{
+		make(map[keyType]decision),
+		make(map[ID]int),
+		make([]Change, 0),
```

Comment:
Please don't initialize empty slices like this - nil slices are perfectly fine.

Or even better, omit the field from initialization:
```golang
cacheStorage{
  decisions: make(...),
  sizes: make(...),
}
```